httpx/router.go

```diff
@@ -0,0 +1,89 @@
+package httpx
+
+import (
+	"fmt"
+	"net/http"
+	"time"
+
+	"github.com/go-chi/chi/v5"
+	"github.com/go-chi/chi/v5/middleware"
+	"github.com/prometheus/client_golang/prometheus/promhttp"
+	"go.uber.org/zap"
+
+	"github.com/zenledger-io/go-utils/httpx/middlewarex"
+	"github.com/zenledger-io/go-utils/version"
+)
+
+var (
+	defaultRequestTimeout = 60 * time.Second
+)
+
+// APIRouterConfig is configuration for an APIRouter.
+type APIRouterConfig struct {
+	// APIVersion is the URL-based API version. All routes created
+	// on a router with this config will be nested under
+	// "/v1" when APIVersion is 1.
+	APIVersion int
+
+	// APIRouter is the router to be mounted under the API verison.
+	APIRouter chi.Router
+
+	// Logger is the base logger to be used for all requests.
+	Logger *zap.Logger
+
+	// RequestTimeout is the maximum amount of time allowed
+	// for a request on the router before processing is stopped.
+	RequestTimeout time.Duration
+
+	// ServiceName is the name of the service running this router.
+	ServiceName string
+}
+
+// NewAPIRouter creates a chi Router that is configured with a set
+// of default middleware and routes for an API. This router reads
+// and writes JSON.
+func NewAPIRouter(cfg APIRouterConfig) chi.Router {
+	r := chi.NewRouter()
+
+	r.Use(middleware.RealIP)
+	r.Use(middleware.RequestID)
+
+	r.Use(middleware.RequestLogger(&middlewarex.ZapLogFormatter{
+		Logger: cfg.Logger,
+	}))
+
+	r.Use(middleware.Recoverer)
+
+	// Request timeouts
+	timeout := cfg.RequestTimeout
+	if timeout == 0 {
+		timeout = defaultRequestTimeout
+	}
+	r.Use(middleware.Timeout(timeout))
```

Comment:
I think it would be good to have a default available but also be able to configure on a per endpoint basis. Some endpoints could be expected to take vastly longer than others

Reply:
Thats a good idea, I'll add that.

Reply:
I started looking into this and I realized that this is really just a sensible default. If specific handlers need longer or shorter timeouts they can ignore the context `Done()` that is passed to them and set their own.

Reply:
The `Done()` of that context is also used for things like shutdown so not sure if ignoring it would be best

Reply:
I think we're saying the same thing. We could use the `Done()` that's provided by default for context cancellation like we should. In the even that we need a different timeout for a specific handler we'd create a new context and use it's `Done()` for cancellation instead.

Reply:
I think the http handlers would really only have access to the request context although things downstream like data sources would have the main context

Reply:
The downstream components should always be using a context that comes from the request. If its a DB connection pool (for example) we'd make sure we called `QueryContext`-like methods and pass the context from the request to these calls.

Only background tasks/goroutines would ever have the main context.

Reply:
Yeah so like in tax calc app, an example would be the `QueryHandler` that we call `Start` on in main gets passed the main context, shutting that down stops all queries. Where as you said, the `QueryContext` function on the handler gets the request context which controls just that query

Reply:
Are you looking for a way to stop all queries? I can't tell if we're on the same page or not.